.github/workflows/provenance.yaml

```diff
@@ -73,3 +76,59 @@ jobs:
       - name: Push provenance branch
         if: github.event_name == 'push'
         run: git -C ./provenance-branch push --follow-tags
+
+      # Cache ent to speed up future runs.
+      # See https://doc.rust-lang.org/nightly/cargo/guide/cargo-home.html#caching-the-cargo-home-in-ci
+      - name: Cache Ent
+        id: cache-ent
+        uses: actions/cache@v3
+        with:
+          path: /usr/local/bin/ent
+          key: ${{ runner.os }}-ent
+
+      # Download the Ent CLI and configure a remote with write access using a private API key, if
+      # available (GitHub secrets are not available on forks, see
+      # https://docs.github.com/en/actions/security-guides/encrypted-secrets#using-encrypted-secrets-in-a-workflow).
+      # If this step is run from a fork, the API key will be empty, and Ent will work in read-only
+      # mode.
+      # See https://github.com/google/ent
+      - name: Download Ent CLI
+        if: steps.cache-ent.outputs.cache-hit != 'true'
+        env:
+          ENT_URL: https://ent-server-62sa4xcfia-ew.a.run.app
+          ENT_DIGEST: sha256:944a34854a2bf9c5d32f3bffa93885ee1c7ef8ab0f4fcc30898a981050ae4233
+        run: |
+          curl --fail ${ENT_URL}/raw/${ENT_DIGEST} > /usr/local/bin/ent
+          chmod +x /usr/local/bin/ent
+          ent
+          cat <<EOF > ~/.config/ent.toml
+          [[remotes]]
+          name = 'ent-store'
+          url = '${ENT_URL}'
+          write = true
+          api_key = '${{ secrets.ENT_API_KEY }}'
+          EOF
+
+      # Upload artifacts to Ent, where they will be retained and publicly accessible by their hash.
+      # This only applies to `push` events, since it needs access to a valid Ent API key (which is
+      # not available on forks / PRs).
+      - name: Upload to Ent
+        if: github.event_name == 'push'
+        run: |
+          ent put ${{inputs.binary_path}}
+
+      # Build reproducibility index on `push` events.
```

Comment:
My expectations was that we could generate the provenance *and* upload to Ent and generate the repro index from the same build config file, since that has all the information necessary. Otherwise, like here, we have to repeat the paths of the built files. Something like another workflow that takes the same build config file as input, then it first runs the provenance generation with the workflow you just added, and then keeps going by uploading the artifact to ent and printing out a provenance diff. I think it should also be possible to control this with an additional input (e.g. `upload_artifacts` and `print_diff` could be added to the workload, and those steps only executed if they are set to true).

Also, my intention is to eventually actually delete the reproducibility index file and corresponding branch, since all the information can be reconstructed from the provenance branch, right? The provenance information is a strict superset of the reproducibiliy information, and anyone should be able to derive something like repro index just from the list of provenances. For instance, from a git commit, we can look up the parent commit id, and then use the provenance branch to find out what the hash of the generated code was at that commit. Then basically we can just show a diff from the previous provenance and the new provenance (perhaps removing noisy fields like timestamps). WDYT?

Reply:
> My expectations was that we could generate the provenance and upload to Ent and generate the repro index from the same build config file, since that has all the information necessary. Otherwise, like here, we have to repeat the paths of the built files. Something like another workflow that takes the same build config file as input, then it first runs the provenance generation with the workflow you just added, and then keeps going by uploading the artifact to ent and printing out a provenance diff. I think it should also be possible to control this with an additional input (e.g. upload_artifacts and print_diff could be added to the workload, and those steps only executed if they are set to true).

I am not entirely sure I understand the approach you are suggesting. Here I am trying to avoid running the build step multiple times, because it takes time and consumes resources. Ideally, the upload to ent and reproducibility index steps could extract the `output_path` from all the files in the `buildconfigs` folder and do the operations for all those files, but currently there are two problems that prevent implementing this solution:
- The build commands generate files that are not included in the build config (e.g., the oak_functions_loader_unsafe)
- There are config files that cannot be built (i.e., the crossvm binary because the build command for it is removed in #3099).

If we can cleanup the builds and fix these issues these workflows will be a lot nicer. 
 
> Also, my intention is to eventually actually delete the reproducibility index file and corresponding branch, since all the information can be reconstructed from the provenance branch, right? The provenance information is a strict superset of the reproducibiliy information, and anyone should be able to derive something like repro index just from the list of provenances. For instance, from a git commit, we can look up the parent commit id, and then use the provenance branch to find out what the hash of the generated code was at that commit. Then basically we can just show a diff from the previous provenance and the new provenance (perhaps removing noisy fields like timestamps). WDYT?

Yes. This is what we eventually want. But do you still want the diff to be published as a comment on the PR? If so, that should be done in a separate PR. If you think the diff should be provided in a separate tool, then I can just remove the reproducibility index in this PR, and we'll build the tool later.

Reply:
> I am not entirely sure I understand the approach you are suggesting. Here I am trying to avoid running the build step multiple times, because it takes time and consumes resources. Ideally, the upload to ent and reproducibility index steps could extract the `output_path` from all the files in the `buildconfigs` folder and do the operations for all those files, but currently there are two problems that prevent implementing this solution:

I was thinking to leave that as an optimisation for later on; for now I would just create an action that can build and upload the result of a single build config file, and run that against all the build config files, one after the other, even if it means that it will repeat some work across them (if that's the case, it seems something that the build process should solve, e.g. by improving its own caching). I don't think we should have another way that uses the build config to build more stuff than actually specified in the target of that build config, and then expect to upload all of them together -- if we really need such a use case then we should design the build config to allow for that, e.g. by specifying multiple outputs, and creating provenance for each of them separately (although I personally prefer the simplicity of just one target per build file, if it's possible and practical). Otherwise this seems to show that we designed a format that does not cover our own needs and we need to work around it, which means other projects will also probably run into similar issues, and we should not require that they also have extra logic that repeats some of the information again in the GitHub actions file. 

Ideally the maintainer of a project that wants to opt in transparent release should just add all the build config files for all their targets, and mechanically add them to the GitHub action that we provide to them.

> * The build commands generate files that are not included in the build config (e.g., the oak_functions_loader_unsafe)

I think we should add a build config that builds that too, right? In fact we should add build configs for all the things we care about building as stand-alone binaries. I imagine build configs similar to top level makefile targets: they define everything necessary to build a given binary. Then, from the GitHub actions config, we just need to invoke the builder action for each of them. Anything else in the GitHub actions config should basically just be running tests / examples.

> * There are config files that cannot be built (i.e., the crossvm binary because the build command for it is removed in [Remove unused xtask options #3099](https://github.com/project-oak/oak/pull/3099)).

Yes that should be reverted, and tests added so that it is not possible to accidentally remove it again -- would adding the build config file and pointing to that from the GitHub actions file enough to make sure the build will fail in that case, so it cannot happen in the future?

> Yes. This is what we eventually want. But do you still want the diff to be published as a comment on the PR? If so, that should be done in a separate PR. If you think the diff should be provided in a separate tool, then I can just remove the reproducibility index in this PR, and we'll build the tool later.

Yes I think eventually we should have the tool (or some other way of looking up a diff between two provenances), and also run that tool to generate the diff from the previous commit, and post that as a comment. The tool may be super simple, for instance given a commit id, it may just download the correct file from the provenance branch, and print it out (then in the GH action we can diff the previous and the new one ourselves).

Reply:
> Ideally the maintainer of a project that wants to opt in transparent release should just add all the build config files for all their targets, and mechanically add them to the GitHub action that we provide to them.

This is true. The problems discussed above are not due to the transparent release, but are due to suboptimal build commands on Oak (the maintainer of the project that has opt into transparent release), and should be fixed by Oak. 



> > * There are config files that cannot be built (i.e., the crossvm binary because the build command for it is removed in [Remove unused xtask options #3099](https://github.com/project-oak/oak/pull/3099)).
> 
> Yes that should be reverted, and tests added so that it is not possible to accidentally remove it again -- would adding the build config file and pointing to that from the GitHub actions file enough to make sure the build will fail in that case, so it cannot happen in the future?

Note that GitHub action already pointed to the config using the removed command. The action was not failing, intentionally, to avoid any potential flakiness in the provenance generation to block merges. The tooling seems stable now, and this PR makes the GitHub action fail. Once #3099 is reverted and the related TODO here is resolved, the workflow will fail and prevent this from happening in future.